src/client.ts

```diff
@@ -569,6 +600,102 @@ in an environmental variable named FAUNA_SECRET or pass it to the Client\
   }
 }
 
+export type StreamEventHandler = (event: StreamEvent) => void;
+
+export class StreamClient {
+  #callbacks: Record<StreamEventType, StreamEventHandler[]> = {
+    start: [],
+    add: [],
+    remove: [],
+    update: [],
+    error: [],
+  };
+  #query: StreamToken;
+  #clientConfiguration: Record<string, any>;
+  #httpStreamClient: HTTPStreamClient;
+  #streamAdapter: StreamAdapter | null = null;
+
+  // TODO: make clientConfiguration and httpStreamClient optional
+  constructor(
+    query: StreamToken,
+    clientConfiguration: StreamClientConfiguration,
+    httpStreamClient: HTTPStreamClient
+  ) {
+    this.#query = query;
+    this.#clientConfiguration = clientConfiguration;
+    this.#httpStreamClient = httpStreamClient;
+  }
+
+  on(type: StreamEventType, callback: StreamEventHandler) {
```

Comment:
we probably should give the option to unsubscribe an existing handler

Reply:
That would be nice. Do you have thoughts on what would be a good way to do that? Note that this implementation allows multiple callbacks for the same event type.

A few ways come to mind for me:

1. `remove(type: StreamEventType)`, removes ALL callbacks for a given event type.
2. Provide a third parameter for options when adding a callback, which could include a name/id for the callback. e.g. `on(type: StreamEventType, callback: StreamEventHandler), options: { id: string })`. Then you could do `stream.remove('callbackName').
3. return a handler to the callback from `on`, instead of the StreamClient. This would mean that you couldn't chain `.on(...)` calls, but you could stash that callback to cancel it later.

```
// option 3
const stream = client.stream(token)
const addHandle = stream.on('add', (e) => ...)

// ... later
addHandle.remove() // not sure what to do with this dangling handler now, though.
```

Reply:
It seems a bit cumbersome to remove a single handler after the stream starts. It might error prone to setup the client, then start the stream, then at some point later a callback stops being called because some other part of the code base removed a handler. If we need to support advanced event handling, I'd prefer a more general event callback where you catch all events instead. Users can write their own complex logic there.

Reply:
At a second look, it might be even confusing to users that we design an API for selecting event based on their types but, they get all events in the wire! One might expect that they are saving on network footprint by only handling some events. Perhaps the more general "catch all" API would make the whole iteration a bit more clear overall.

Reply:
> Perhaps the more general "catch all" API would make the whole iteration a bit more clear overall.

Do you mean the async-iterator API? e.g. `for await (const event in stream.iter())`?

Reply:
Idk which API shape is the most idiomatic. My point is: having the driver select event types may confuse users into thinking they are only getting those events.

The async-iterator API as is right now aligns with my perspective since it does clearly informs users that they are getting all events. The callback based API would have to be reshaped into `.on(event => {..})` to align with what I'm describing here.

Reply:
feel free to correct me if I'm wrong but I think he is referring to something like the following:
```
const stream = client.stream(stream)
      .onEvent((type, data) => {
          switch (type) {
          case 'update':
              console.log(`New doc's data ${data}`)
              break
          case 'remove':
              console.log('Doc deleted')
              break
          }
      })
      .onError(error => {
          console.log(error)
      })
```
though I suppose in that example we had the error case special handled.  could move that into the general event handling as well though I don't think it is necessarily the worst to handle errors in that way.

Reply:
I don't think it's a problem that all events are being triggered even if you don't register a callback for them. This should be familiar to JS devs. This is how v4 driver works for reference, not that that means we _should_ do it that way.

Should registering a callback overwrite the existing callback, or should it be added to a list?

If we break up the `on` method to have multiple such as `onStart`, `onEvent`/`onData` and `onError` methods, then how do you want to do de-registering events?

Reply:
FWIW, if we are going to only have one method between callbacks and async iterator, then my vote is currently on async iterator. Then this becomes moot, and, to Chase's question on error handling, it is clear where any network errors would be caught and handled.